src/System.Buffers/src/System/Buffers/DefaultArrayPool.cs

```diff
@@ -0,0 +1,151 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Reflection;
+using System.Runtime.InteropServices;
+using System.Threading;
+
+namespace System.Buffers
+{
+    internal sealed class DefaultArrayPool<T> : ArrayPool<T>
+    {
+        private const int MinimiumArraySize = 16;
+        private int _maxArrayLength;
+        private int _numberOfArraysPerBucket;
+        private bool _isTObjectType;    // cache this to only do the reflection call once
```

Comment:
> I can't take advantage of that yet

I provided an example in that issue of how it could be implemented outside of the runtime.  If it's important that we have that behavior, we could do it that way now and then change to a better implementation when it's available.

I'm not actually sure this is better than nothing, as if we automatically clear it may lead people into a false sense of security that they don't need to specify clearing because the implementation will do it for them.

Seems like it should be an all-or-nothing thing... either we clear any time it's possible there's a reference, or we fully respect the value supplied by the caller.  For simplicity, for now, I'd suggest the latter.  @KrzysztofCwalina may have a different opinion.